File: clients/kotlin_graal/src/main/kotlin/ai_cup_22/model/UnitOrder.kt
package ai_cup_22.model

import ai_cup_22.util.StreamUtil

/**
 * Order for specific unit
 */
class UnitOrder {
    /**
     * Target moving velocity
     */
    var targetVelocity: ai_cup_22.model.Vec2
    /**
     * Target view direction (vector length doesn't matter)
     */
    var targetDirection: ai_cup_22.model.Vec2
    /**
     * Order to perform an action, or None
     */
    var action: ai_cup_22.model.ActionOrder?

    constructor(targetVelocity: ai_cup_22.model.Vec2, targetDirection: ai_cup_22.model.Vec2, action: ai_cup_22.model.ActionOrder?) {
        this.targetVelocity = targetVelocity
        this.targetDirection = targetDirection
        this.action = action
    }

    /**
     * Write UnitOrder to output stream
     */
    @Throws(java.io.IOException::class)
    fun writeTo(stream: java.io.OutputStream) {
        targetVelocity.writeTo(stream)
        targetDirection.writeTo(stream)
        val actionValue = action
        if (actionValue == null) {
            StreamUtil.writeBoolean(stream, false)
        } else {
            StreamUtil.writeBoolean(stream, true)
            actionValue.writeTo(stream)
        }
    }

    /**
     * Get string representation of UnitOrder
     */
    override fun toString(): String {
        var stringBuilder = StringBuilder("UnitOrder { ")
        stringBuilder.append("targetVelocity: ")
        stringBuilder.append(targetVelocity)
        stringBuilder.append(", ")
        stringBuilder.append("targetDirection: ")
        stringBuilder.append(targetDirection)
        stringBuilder.append(", ")
        stringBuilder.append("action: ")
        stringBuilder.append(action)
        stringBuilder.append(" }")
        return stringBuilder.toString()
    }

    companion object {
        /**
         * Read UnitOrder from input stream
         */
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): UnitOrder {
            var targetVelocity: ai_cup_22.model.Vec2
            targetVelocity = ai_cup_22.model.Vec2.readFrom(stream)
            var targetDirection: ai_cup_22.model.Vec2
            targetDirection = ai_cup_22.model.Vec2.readFrom(stream)
            var action: ai_cup_22.model.ActionOrder?
            if (StreamUtil.readBoolean(stream)) {
                action = ai_cup_22.model.ActionOrder.readFrom(stream)
            } else {
                action = null
            }
            return UnitOrder(targetVelocity, targetDirection, action)
        }
    }
}

File: clients/swift/Sources/AiCup22/Model/UnitOrder.swift
/// Order for specific unit
public struct UnitOrder {
    /// Target moving velocity
    let targetVelocity: Vec2

    /// Target view direction (vector length doesn't matter)
    let targetDirection: Vec2

    /// Order to perform an action, or None
    let action: ActionOrder?

    /// Read UnitOrder from input stream
    static func readFrom<S: InputStream>(_ stream: S) -> UnitOrder {
        var targetVelocity: Vec2
        targetVelocity = Vec2.readFrom(stream)
        var targetDirection: Vec2
        targetDirection = Vec2.readFrom(stream)
        var action: ActionOrder?
        if stream.readBool() {
            action = ActionOrder.readFrom(stream)
        } else {
            action = nil
        }
        return UnitOrder(targetVelocity: targetVelocity, targetDirection: targetDirection, action: action)
    }

    /// Write UnitOrder to output stream
    func writeTo<S: OutputStream>(_ stream: S) {
        targetVelocity.writeTo(stream)
        targetDirection.writeTo(stream)
        if action == nil {
            stream.writeBool(false)
        } else {
            stream.writeBool(true)
            let actionValue = action!
            actionValue.writeTo(stream)
        }
    }
}
